python/src/iceberg/catalog/base.py

```diff
@@ -0,0 +1,222 @@
+#  Licensed to the Apache Software Foundation (ASF) under one
+#  or more contributor license agreements.  See the NOTICE file
+#  distributed with this work for additional information
+#  regarding copyright ownership.  The ASF licenses this file
+#  to you under the Apache License, Version 2.0 (the
+#  "License"); you may not use this file except in compliance
+#  with the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing,
+#  software distributed under the License is distributed on an
+#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+#  KIND, either express or implied.  See the License for the
+#  specific language governing permissions and limitations
+#  under the License.
+
+from abc import ABC, abstractmethod
+from typing import Dict, List, Optional, Tuple
+
+from iceberg.table.base import Table, TableSpec
+
+
+class Catalog(ABC):
+    """Base Catalog for table operations like - create, drop, load, list and others.
+
+    Attributes:
+        name(str): Name of the catalog
+        properties(Dict[str, str]): Catalog properties
+    """
+
+    def __init__(self, name: str, properties: Dict[str, str]):
+        self._name = name
+        self._properties = properties
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def properties(self) -> Dict[str, str]:
+        return self._properties
+
+    @abstractmethod
+    def create_table(self, table_spec: TableSpec) -> Table:
+        """Create a table
+
+        Args:
+            table_spec: A specification to create a table
+
+        Returns:
+            Table: the created table instance
+
+        Raises:
+            AlreadyExistsError: If a table with the name already exists
+        """
+
+    @abstractmethod
+    def table(self, namespace: str, name: str) -> Table:
```

Comment:
`namespace` here and throughout the other signatures should probably be a tuple (`Tuple[str, ...]`) as opposed to a string.  After talking with @fokko, I think that's probably a more concise representation of the namespace without having to worry about parsing levels.